File: speedviewlib/src/main/java/com/github/anastr/speedviewlib/components/Indicators/NeedleIndicator.java
package com.github.anastr.speedviewlib.components.Indicators;

import android.content.Context;
import android.graphics.BlurMaskFilter;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RectF;

/**
 * this Library build By Anas Altair
 * see it on <a href="https://github.com/anastr/SpeedView">GitHub</a>
 */

public class NeedleIndicator extends Indicator {

    private Path indicatorPath = new Path();
    private Path circlePath = new Path();
    private Paint circlePaint =  new Paint(Paint.ANTI_ALIAS_FLAG);
    private float bottomY;

    public NeedleIndicator(Context context) {
        super(context);
        circlePaint.setStyle(Paint.Style.STROKE);
        updateIndicator();
    }

    @Override
    protected float getDefaultIndicatorWidth() {
        return dpTOpx(12f);
    }

    @Override
    public float getBottom() {
        return bottomY;
    }

    @Override
    public void draw(Canvas canvas, float degree) {
        canvas.save();
        canvas.rotate(90f + degree, getCenterX(), getCenterY());
        canvas.drawPath(indicatorPath, indicatorPaint);
        canvas.drawPath(circlePath, circlePaint);
        canvas.restore();
    }

    @Override
    protected void updateIndicator() {
        indicatorPath.reset();
        circlePath.reset();
        indicatorPath.moveTo(getCenterX(), getPadding());
        bottomY = (float) (getIndicatorWidth() * Math.sin(Math.toRadians(260))) + getViewSize()*.5f + getPadding();
        float xLeft = (float) (getIndicatorWidth() * Math.cos(Math.toRadians(260))) + getViewSize()*.5f + getPadding();
        indicatorPath.lineTo(xLeft, bottomY);
        RectF rectF = new RectF(getCenterX() - getIndicatorWidth(), getCenterY() - getIndicatorWidth()
                , getCenterX() + getIndicatorWidth(), getCenterY() + getIndicatorWidth());
        indicatorPath.arcTo(rectF, 260, 20f);

        float circleWidth = getIndicatorWidth() *.25f;
        circlePath.addCircle(getCenterX(), getCenterY(), getIndicatorWidth() - circleWidth*.5f +.6f, Path.Direction.CW);

        indicatorPaint.setColor(getIndicatorColor());
        circlePaint.setColor(getIndicatorColor());
        circlePaint.setStrokeWidth(circleWidth);
    }

    @Override
    protected void setWithEffects(boolean withEffects) {
        if (withEffects && !isInEditMode())
            indicatorPaint.setMaskFilter(new BlurMaskFilter(15, BlurMaskFilter.Blur.SOLID));
        else
            indicatorPaint.setMaskFilter(null);
    }
}

File: speedviewlib/src/main/java/com/github/anastr/speedviewlib/components/indicators/NeedleIndicator.kt
package com.github.anastr.speedviewlib.components.indicators

import android.content.Context
import android.graphics.*
import kotlin.math.cos
import kotlin.math.sin

/**
 * this Library build By Anas Altair
 * see it on [GitHub](https://github.com/anastr/SpeedView)
 */

class NeedleIndicator(context: Context) : Indicator<NeedleIndicator>(context) {

    private val indicatorPath = Path()
    private val circlePath = Path()
    private val circlePaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var bottomY: Float = 0.toFloat()

    init {
        width = dpTOpx(12f)
        circlePaint.style = Paint.Style.STROKE
    }

    override fun getBottom(): Float {
        return bottomY
    }

    override fun draw(canvas: Canvas) {
        canvas.drawPath(indicatorPath, indicatorPaint)
        canvas.drawPath(circlePath, circlePaint)
    }

    override fun updateIndicator() {
        indicatorPath.reset()
        circlePath.reset()
        indicatorPath.moveTo(getCenterX(), speedometer!!.padding.toFloat())
        bottomY = (width * sin(Math.toRadians(260.0))).toFloat() + getViewSize() * .5f + speedometer!!.padding.toFloat()
        val xLeft = (width * cos(Math.toRadians(260.0))).toFloat() + getViewSize() * .5f + speedometer!!.padding.toFloat()
        indicatorPath.lineTo(xLeft, bottomY)
        val rectF = RectF(getCenterX() - width, getCenterY() - width, getCenterX() + width, getCenterY() + width)
        indicatorPath.arcTo(rectF, 260f, 20f)

        val circleWidth = width * .25f
        circlePath.addCircle(getCenterX(), getCenterY(), width - circleWidth * .5f + .6f, Path.Direction.CW)

        indicatorPaint.color = color
        circlePaint.color = color
        circlePaint.strokeWidth = circleWidth
    }

    override fun setWithEffects(withEffects: Boolean) {
        if (withEffects && !speedometer!!.isInEditMode)
            indicatorPaint.maskFilter = BlurMaskFilter(15f, BlurMaskFilter.Blur.SOLID)
        else
            indicatorPaint.maskFilter = null
    }
}
